docs/specification/target.md

```diff
@@ -0,0 +1,97 @@
+
+struct OutControl {
+    PortId outputPort;
+}
+
+const PortId DROP_PORT = 0xF;
+const PortId CPU_OUT_PORT = 0xE;
+const PortId RECIRCULATE_OUT_PORT = 0xD;
+
+parser Parser<H>(packet_in b, out H parsedHeaders);
+
+control Pipe<H>(inout H headers,
+                in error parseError,
+                in InControl inCtrl,
+                out OutControl outCtrl);
+
+control Deparser<H>(inout H outputHeaders, packet_out b);
+
+package VSS<H>(Parser<H> p,
+               Pipe<H> map,
+               Deparser<H> d);
+
+extern Checksum16 {
+    Checksum16();
+    void clear();
+    void update<T>(in T data);
+    void remove<T>(in T data);
+    bit<16> get();
+}
+```
+
+First, the user must provide the type `obj`, which will be used to represent target-provided stateful value. In the case of `VSS`, only one such data structure is needed -- the underlying value of the `Checksum16` extern. Some other commonly used externs include counter and register arrays, as in `V1 Model`. The types `state` and `extern` are parameterized on the type `obj` and should be copied into the implementation as they appear in the signature.
+
+The user must also provide the functions `write_header_field` and `read_header_field`. These functions will be called by the main interpreter when reading and writing to headers, and they are intended to capture the fact that the semantics in these cases is left target-dependent by the P4 specification. However, the abstraction in its current form is not necessarily expressive enough to capture all possible decisions for header reads and writes. The average user will likely want to reuse our implementations of these functions in `v1model.ml` and `ebpf.ml`.
+
+The next required value is the `eval_extern` function. This function takes as its arguments the name of the extern to evaluate, the envrionment and state, the type arguments of the extern call (e.g. a concrete value for `T` in the case of `update` and `remove` in `VSS` provided by the type checker), and the arguments paired with their types. For implementation, we may assume that the arguments are provided in the correct order and that the list is the proper length. The is also the place where we use the number and types of the arguments to distinguish between different externs of the same name, as permitted by the P4 specification. Note that in the case of an invocation of an extern function using dot-notation, e.g. `checksum.clear()`, the value of `checksum` will be available as the additional first argument in the list of arguments. The extern evaluation should then return as a tuple the updated environment (though most externs do not change the environment), the update state, a signal (almost always `Continue`), and the return value. Implementing the externs will require some degree of familiarity with our types for values, environments, and states. Also, note that any mutation of an extern object should be done by updating the mapping in the state via `insert_extern`.
```

Comment:
In the sentence:
>Note that in the case of an invocation of an extern function using dot-notation, e.g. `checksum.clear()`, the value of `checksum` will be available as the additional first argument in the list of arguments.

I was confused as to which argument of `eval_extern` `checksum` corresponds to. I believe it roughly corresponds to `(value * Type.t) list`, but I don't understand exactly how...

I think it would be really helpful to have named arguments in `target.mli` as well as `OCamldoc` documentation to explain what the different arguments are and how they are used.